File: May2022/Java/UniquePathsii.java
public class Solution {
	// TC : O(n*m)
	// SC : O(n*m)
	/**
	 * @param obstacleGrid: A list of lists of integers
	 * @return: An integer
	 */
	public int uniquePathsWithObstacles(int[][] obstacleGrid) {
		// write your code here
		int m = obstacleGrid.length, n= obstacleGrid[0].length;
		int[][] dp = new int[m][n];

		for(int i=0;i<n;i++){
			if(obstacleGrid[0][i] == 1){
				break;
			} else {
				dp[0][i] =1;
			}
		}
		// m rows
		for(int i=0;i<m;i++){
			if(obstacleGrid[i][0]==1){
				break;
			}
			else{
				dp[i][0] =1;
			}
		}

		for(int i=1;i<m;i++){
			for(int j=1;j<n;j++){
				if(obstacleGrid[i][j]==1){
					dp[i][j] =0;
				} else {
					dp[i][j] = dp[i-1][j] + dp[i][j-1];
				}
			}
		}



		return dp[m-1][n-1];
	}
}

File: August2023 / Unique Paths II.cpp
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& grid) {
        vector<vector<int>> visited(100,vector<int>(100,-1));
        return dfs(grid,visited,0,0);
    }

    int dfs(vector<vector<int>>& grid,vector<vector<int>>& visited,int i,int j){
        if(i==grid.size() || j==grid[0].size() || grid[i][j]==1) return 0;
        if(i==grid.size()-1 && j==grid[0].size()-1) return 1; 
        if(visited[i][j]!=-1) return visited[i][j];
        visited[i][j]=dfs(grid,visited,i,j+1)+dfs(grid,visited,i+1,j);
        return visited[i][j];
    }
};